File: app/src/main/java/web/id/berviantoleo/rgbcounter/MainActivity.java
/**
 * Copyright 2018 Bervianto Leo Pratama
 * <p>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package web.id.berviantoleo.rgbcounter;

import android.app.AlertDialog;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;

import androidx.appcompat.app.AppCompatActivity;

import com.esafirm.imagepicker.features.ImagePicker;
import com.esafirm.imagepicker.features.ReturnMode;
import com.esafirm.imagepicker.model.Image;
import com.facebook.drawee.view.SimpleDraweeView;
import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import dmax.dialog.SpotsDialog;

public class MainActivity extends AppCompatActivity {

    @BindView(R.id.image_holder)
    protected SimpleDraweeView imageHolder;

    @BindView(R.id.chart)
    protected LineChart lineChart;

    private LineData lineData;
    private AlertDialog dialog;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        dialog = new SpotsDialog.Builder().setContext(this).build();
        lineData = new LineData();
    }

    @OnClick(R.id.image_holder)
    protected void clickImage() {
        ImagePicker.create(this)
                .returnMode(ReturnMode.ALL)
                .folderMode(true) // folder mode (false by default)
                .toolbarFolderTitle("Select Folder") // folder selection title
                .toolbarImageTitle("Select Image") // image selection title
                .single() // single mode
                .showCamera(true) // show camera or not (true by default)
                .imageDirectory("RGB_Counter") // directory name for captured image  ("Camera" folder by default)
                .enableLog(false) // disabling log
                .start(); // start image picker activity with request code
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (ImagePicker.shouldHandle(requestCode, resultCode, data)) {
            Image image = ImagePicker.getFirstImageOrNull(data);
            if (image != null) {
                String path = image.getPath();
                Uri uri = Uri.parse("file://" + path);
                imageHolder.setImageURI(uri);
                Bitmap bitmap = BitmapFactory.decodeFile(path);
                new CountColour(this).execute(bitmap);
            }
        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    private static class CountColour extends AsyncTask<Bitmap, Void, Void> {

        private final WeakReference<MainActivity> activityWeakReference;
        private long start;

        CountColour(MainActivity context) {
            activityWeakReference = new WeakReference<>(context);
        }

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            MainActivity activity = activityWeakReference.get();
            activity.lineData.clearValues();
            activity.lineData.notifyDataChanged();
            activity.dialog.show();
            start = System.nanoTime();
        }

        @Override
        protected Void doInBackground(Bitmap... params) {
            MainActivity activity = activityWeakReference.get();
            Bitmap bitmap = params[0];
            int width = bitmap.getWidth();
            int height = bitmap.getHeight();
            int[] red = new int[256];
            int[] green = new int[256];
            int[] blue = new int[256];
            for (int j = 0; j < height; j++) {
                for (int i = 0; i < width; i++) {
                    int color = bitmap.getPixel(i, j);
                    int redColor = (color >> 16) & 0xff;
                    int greenColor = (color >> 8) & 0xff;
                    int blueColor = color & 0xff;
                    red[redColor] += 1;
                    green[greenColor] += 1;
                    blue[blueColor] += 1;
                }
            }
            List<Entry> entriesRed = new ArrayList<>();
            List<Entry> entriesGreen = new ArrayList<>();
            List<Entry> entriesBlue = new ArrayList<>();
            for (int i = 0; i < 256; i++) {
                entriesRed.add(new Entry(i, red[i]));
                entriesGreen.add(new Entry(i, green[i]));
                entriesBlue.add(new Entry(i, blue[i]));
            }
            LineDataSet dataSetRed = new LineDataSet(entriesRed, "red");
            LineDataSet dataSetGreen = new LineDataSet(entriesGreen, "green");
            LineDataSet dataSetBlue = new LineDataSet(entriesBlue, "blue");
            dataSetRed.setColor(Color.RED);
            dataSetRed.setDrawCircles(false);
            dataSetGreen.setColor(Color.GREEN);
            dataSetGreen.setDrawCircles(false);
            dataSetBlue.setColor(Color.BLUE);
            dataSetBlue.setDrawCircles(false);
            activity.lineData.addDataSet(dataSetRed);
            activity.lineData.addDataSet(dataSetGreen);
            activity.lineData.addDataSet(dataSetBlue);
            return null;
        }

        @Override
        protected void onPostExecute(Void aVoid) {
            MainActivity activity = activityWeakReference.get();
            activity.lineChart.setData(activity.lineData);
            activity.lineChart.invalidate();
            activity.dialog.dismiss();
            long end = System.nanoTime();
            long duration = end - start;
            Log.i("Process Photo", String.format("Waktu dibutuhkan : %d", duration));
        }
    }
}

File: app/src/main/java/web/id/berviantoleo/rgbcounter/MainActivity.kt
/**
 * Copyright 2018 Bervianto Leo Pratama
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package web.id.berviantoleo.rgbcounter

import android.app.AlertDialog
import android.content.ContentValues
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.MediaStore
import android.util.Log
import android.widget.Toast
import androidx.activity.result.PickVisualMediaRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import dmax.dialog.SpotsDialog
import web.id.berviantoleo.rgbcounter.databinding.ActivityMainBinding
import java.io.OutputStream
import java.lang.ref.WeakReference


class MainActivity : AppCompatActivity() {
    private val lineData: LineData = LineData()
    private lateinit var dialog: AlertDialog
    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view = binding.root
        setContentView(view)
        val launcher = registerForActivityResult(ActivityResultContracts.PickVisualMedia()) { uri ->
            if (uri != null) {
                binding.imageHolder.setImageURI(uri, this)
                val runnableCounter = ColourCounter(this, uri)
                Thread(runnableCounter).start()
            } else {
                Log.d("PhotoPicker", "No media selected")
                Toast.makeText(this, "No media selected", Toast.LENGTH_SHORT).show()
            }
        }
        dialog = SpotsDialog.Builder().setContext(this).build()
        binding.imageHolder.setOnClickListener {
            launcher.launch(PickVisualMediaRequest(ActivityResultContracts.PickVisualMedia.ImageOnly))
        }
        binding.saveToGallery.setOnClickListener {
            val fileName = "chart-${System.currentTimeMillis()}.jpg"
            saveChartToGallery(fileName)
        }
    }

    class ColourCounter(context: MainActivity, private val uri: Uri) : Runnable {
        private val activityWeakReference: WeakReference<MainActivity> = WeakReference(context)
        private var start: Long = 0

        override fun run() {
            val activity = activityWeakReference.get() ?: return
            activity.runOnUiThread {
                activity.lineData.clearValues()
                activity.lineData.notifyDataChanged()
                activity.dialog.show()
            }

            val bitmap = activity.decodeBitmap(uri)
            if (bitmap == null) {
                activity.runOnUiThread {
                    activity.dialog.dismiss()
                    Toast.makeText(activity, "Failed to decode image", Toast.LENGTH_SHORT).show()
                }
                return
            }

            start = System.nanoTime()
            val width = bitmap.width
            val height = bitmap.height
            val red = IntArray(256)
            val green = IntArray(256)
            val blue = IntArray(256)

            val pixels = IntArray(width * height)
            bitmap.getPixels(pixels, 0, width, 0, 0, width, height)
            bitmap.recycle()

            for (pixel in pixels) {
                val r = (pixel shr 16) and 0xFF
                val g = (pixel shr 8) and 0xFF
                val b = pixel and 0xFF
                red[r]++
                green[g]++
                blue[b]++
            }

            val entriesRed: MutableList<Entry> = ArrayList()
            val entriesGreen: MutableList<Entry> = ArrayList()
            val entriesBlue: MutableList<Entry> = ArrayList()
            for (i in 0..255) {
                entriesRed.add(Entry(i.toFloat(), red[i].toFloat()))
                entriesGreen.add(Entry(i.toFloat(), green[i].toFloat()))
                entriesBlue.add(Entry(i.toFloat(), blue[i].toFloat()))
            }
            val dataSetRed = LineDataSet(entriesRed, "red")
            val dataSetGreen = LineDataSet(entriesGreen, "green")
            val dataSetBlue = LineDataSet(entriesBlue, "blue")
            dataSetRed.color = Color.RED
            dataSetRed.setDrawCircles(false)
            dataSetGreen.color = Color.GREEN
            dataSetGreen.setDrawCircles(false)
            dataSetBlue.color = Color.BLUE
            dataSetBlue.setDrawCircles(false)

            activity.runOnUiThread {
                activity.lineData.addDataSet(dataSetRed)
                activity.lineData.addDataSet(dataSetGreen)
                activity.lineData.addDataSet(dataSetBlue)
                
                activity.binding.chart.data = activity.lineData
                activity.binding.chart.invalidate()
                activity.dialog.dismiss()
                
                if (activity.lineData.dataSetCount > 0) {
                    val end = System.nanoTime()
                    val duration = end - start
                    val timeProcess = "Time to process: ${duration / 1_000_000} ms"
                    Log.i("Process Photo", timeProcess)
                    Toast.makeText(activity, timeProcess, Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(activity, "No data to display", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    private fun saveChartToGallery(fileName: String) {
        val bitmap = binding.chart.chartBitmap
        val contentValues = ContentValues().apply {
            put(MediaStore.MediaColumns.DISPLAY_NAME, fileName)
            put(MediaStore.MediaColumns.MIME_TYPE, "image/jpeg")
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                put(MediaStore.MediaColumns.RELATIVE_PATH, "Pictures/RGBCounter")
                put(MediaStore.MediaColumns.IS_PENDING, 1)
            }
        }

        val uri = contentResolver.insert(MediaStore.Images.Media.EXTERNAL_CONTENT_URI, contentValues)
        if (uri == null) {
            Toast.makeText(this, "Failed to create MediaStore entry", Toast.LENGTH_SHORT).show()
            return
        }

        try {
            contentResolver.openOutputStream(uri)?.use { outputStream ->
                bitmap.compress(Bitmap.CompressFormat.JPEG, 90, outputStream)
            }

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                contentValues.clear()
                contentValues.put(MediaStore.MediaColumns.IS_PENDING, 0)
                contentResolver.update(uri, contentValues, null, null)
            }
            Toast.makeText(this, "Saved to Gallery", Toast.LENGTH_SHORT).show()
        } catch (e: Exception) {
            Log.e("MainActivity", "Error saving chart", e)
            contentResolver.delete(uri, null, null)
            Toast.makeText(this, "Failed to save chart: ${e.message}", Toast.LENGTH_SHORT).show()
        }
    }

    private fun decodeBitmap(uri: Uri): Bitmap? {
        return try {
            val options = BitmapFactory.Options().apply {
                inJustDecodeBounds = true
            }
            contentResolver.openInputStream(uri)?.use {
                BitmapFactory.decodeStream(it, null, options)
            }

            var sampleSize = 1
            while (options.outWidth / (sampleSize * 2) >= 1024 && options.outHeight / (sampleSize * 2) >= 1024) {
                sampleSize *= 2
            }

            val decodeOptions = BitmapFactory.Options().apply {
                inSampleSize = sampleSize
            }
            contentResolver.openInputStream(uri)?.use {
                BitmapFactory.decodeStream(it, null, decodeOptions)
            }
        } catch (e: Exception) {
            Log.e("ColourCounter", "Error decoding bitmap", e)
            null
        }
    }
}